napari/_qt/qt_chunk_receiver.py

```diff
@@ -0,0 +1,55 @@
+from qtpy.QtCore import QObject, Signal
+
+from ..layers.base import Layer
+from ..utils.chunk import ChunkRequest, chunk_loader
+from ..utils.perf import add_instant_event
+
+
+class QtChunkReceiver(QObject):
+    """Shuttles newly loaded chunks from the ChunkLoader to their layers.
+
+    Notes
+    -----
+    The ChunkLoader's chunk_loaded event might be signaled in a worker
+    thread. The documentation only guarantees it will be signaled in thread
+    in this process, in theory could be any thread.
+
+    We do not want to call into model/layer code in a worker thread so we
+    signal ourselves in the GUI thread. If we are already in the GUI thread
+    this does no harm and happens instantly.
+    """
+
+    chunk_loaded_gui = Signal(Layer, ChunkRequest)
+
+    def __init__(self):
+        super().__init__()
+
+        # ChunkLoader signals us when a chunk has been loaded.
+        chunk_loader.events.chunk_loaded.connect(self._chunk_loaded_worker)
+
+        # We signal ourselves in the GUI thread.
```

Comment:
The problem is `ChunkLoader._done` can be called in a worker thread. But we want to pass chunks to layers in the GUI thread. 

We use `future.add_done_callback()` to set `ChunkLoader._done` as the callback for our `ChunkLoader` futures. The concurrent.futures documentation states that `_done()` will be called "in a thread belonging to the process that added them". It purposely does not say which thread. Presumably so the implementation can decide.

Under MacOS the done callback seems to always be run in the worker thread. In Windows the worker thread could just call `PostMessage` with a pointer to the chunk, and the GUI thread would receive the message. So it would be trivial.

With PyQT/Qt I did not see a simple way to do it. The `QtChunkReceiver` is basically a big workaround for the fact that `ChunkLoader._done()` cannot "call into" the GUI thread.

So right now `ChunkLoader._done()` thread signals the `chunk_loaded` event. `QtChunkReceiver` listens to `chunk_loaded` and uses Signal/Slots to call `_chunk_loaded_gui` in the GUI thread.

Basically we are just wiring our "event" to call into a Qt slot. So that from a worker thread we can cause code to run in the GUI thread. I think we could pretty easily make `QtChunkReceiver` could probably be something generic like a `EventReflectorGUI`. For chunks we tell `EventReflectorGUI` what event to listen for, and what code to run in the GUI thread.

I didn't want to do that yet because the whole thing seems a like a big workaround and kind of hack. But if we decide there is no better way, making it generic would be gone, so no one else has to invent this thing.

Reply:
Note that if `_done()` **is** called in the GUI thread, for some reason, it should still work fine. The Signal should just immediately call the Slot. As I understand it Signal basically says "if I'm in the GUI thread then immediately and directly call the Slot, otherwise post a Qt Event and the GUI thread will call the Slot when its next processing events." Which his what we want.

Reply:
I think I understand the purpose of `QtChunkReceiver`... and the goal of reflecting a signal into a different thread.  but I still dont see A) how a method of a `QtChunkReceiver` object would ever be running in another thread (`ChunkLoader._done` sure... but how does this chunkReceiver get there), and B) how an object calling its own signal is helping here.  For instance, have you found a scenario in which removing `chunk_loaded_gui = Signal(Layer, ChunkRequest)`and using this instead doesn't work?:

```python
    def _chunk_loaded_worker(self, event) -> None:
        add_instant_event("_chunk_loaded_worker")
        # self.chunk_loaded_gui.emit(event.layer, event.request) # not this
        self._chunk_loaded_gui(event.layer, event.request)
```

(for me, making that switch works fine... but I'm probably missing an edge case?)

Basically, I'm still confused as to how the `QtChunkReceiver` receiver instance would ever _itself_ need cross thread communication, which is the only reason I can see for having an object listen to its own event like this, rather than just directly calling its own method.

Reply:
`ChunkLoader._done` signals an event which `QtChunkReceiver` listens to. If `ChunkLoader._done` is in a worker thread then `QtChunkReciever._chunk_loaded_worker` will be called in a worker thread. That's why I put the `worker` suffix on there and included the comment "The method is probably running in a worker thread.".

One good way to confirm this is place breakpoint in the `QtChunkReciever._chunk_loaded_worker` function. When then breakpoint hits look at which thread you are in. For me on MacOS it's always in a worker thread.

The only reason `QtChunkReceiver` exists is so the `ChunkLoader` worker can signal to the GUI thread that a chunk was loaded. I could not find any other way to do that. Play around the debugger and play close attention to which thread you are in. If you can find a better way I'd love to get rid of `QtChunkReceiver` entirely.

As I suggest elsewhere the obvious improvement here is turn `QtChunkReceiver` into something generic like `QtEventReflector` and then create an instance of that which does the chunkloaded.  But if we can avoid having a `QtChunkLoader` and a `QtEventReflector` all together that would be completely awseome.

As I said elsewhere this 60 line file would be 1 line with Windows. So I'm not sure how Qt engineering themselves into this hole. Which suggests maybe there is another way?